### PR TITLE
Swap `InfiniteDataloader` for `persistent_workers=True`

### LINKED ISSUE #6741: InfiniteDataLoader does not shuffle at each epoch
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Training

### Bug

It looks like the `InfiniteDataLoader` does not reshuffle the data when used with `num_workers!=0`. A small test program to prove my point:
```python
import torch.distributed as dist
from torch.utils.data.dataloader import DataLoader
from torch.utils.data.distributed import DistributedSampler
from utils.datasets import InfiniteDataLoader


def main(args):
    dist.init_process_group(backend="nccl" if dist.is_nccl_available() else "gloo")
    if dist.get_rank() == 0:
        print(args)

    dataset = list(range(16))
    sampler = DistributedSampler(dataset, shuffle=True)

    if args.infinite:
        loader = InfiniteDataLoader(dataset, batch_size=4, sampler=sampler, num_workers=args.num_workers, pin_memory=True)
    else:
        loader = DataLoader(dataset, batch_size=4, sampler=sampler, num_workers=args.num_workers, pin_memory=True)

    batches = {}

    for epoch in range(3):
        sampler.set_epoch(epoch)
        batches[epoch] = []
        for batch in enumerate(loader):
            batches[epoch].append(batch)

    if dist.get_rank() == 0:
        for epoch, epoch_bacthes in batches.items():
            for i, batch in enumerate(epoch_bacthes):
                print(f'rank: {dist.get_rank()}, epoch {epoch}, batch [{i:02d}] {batch}')

    dist.barrier()

    if dist.get_rank() != 0:
        for epoch, epoch_bacthes in batches.items():
            for i, batch in enumerate(epoch_bacthes):
                print(f'rank: {dist.get_rank()}, epoch {epoch}, batch [{i:02d}] {batch}')

    dist.barrier()


if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('--num-workers', type=int, default=4)
    parser.add_argument('--infinite', action='store_true', default=False)

    args = parser.parse_args()
    main(args)
```

Below is the output of different ways of calling the test program. If it is called with `--infinite` and `--num-workers!=0` every epoch has the same batches. Note how only the first case exhibits this behavior.

Since `InfiniteDataLoader` was introduced to avoid recreating the workers at every epoch, maybe a simple fix would be to use `DataLoader` with `persistent_workers=True`? Unless having the same order of shuffling on each epoch is what was intended?

When called like this:
```
CUDA_VISIBLE_DEVICES=2,3 python3 -m torch.distributed.run --nproc_per_node 2 test_distributed.py --num-workers=4 --infinite
```
it outputs:
```
Namespace(infinite=True, num_workers=4)
rank: 0, epoch 0, batch [00] (0, tensor([12,  9, 11, 13]))
rank: 0, epoch 0, batch [01] (1, tensor([ 2, 15,  4,  7]))
rank: 0, epoch 1, batch [00] (0, tensor([12,  9, 11, 13]))
rank: 0, epoch 1, batch [01] (1, tensor([ 2, 15,  4,  7]))
rank: 0, epoch 2, batch [00] (0, tensor([12,  9, 11, 13]))
rank: 0, epoch 2, batch [01] (1, tensor([ 2, 15,  4,  7]))
rank: 1, epoch 0, batch [00] (0, tensor([10,  6,  8,  5]))
rank: 1, epoch 0, batch [01] (1, tensor([14,  0,  3,  1]))
rank: 1, epoch 1, batch [00] (0, tensor([10,  6,  8,  5]))
rank: 1, epoch 1, batch [01] (1, tensor([14,  0,  3,  1]))
rank: 1, epoch 2, batch [00] (0, tensor([10,  6,  8,  5]))
rank: 1, epoch 2, batch [01] (1, tensor([14,  0,  3,  1]))
```

When called like this:
```
CUDA_VISIBLE_DEVICES=2,3 python3 -m torch.distributed.run --nproc_per_node 2 test_distributed.py --num-workers=0 --infinite
```
it outputs:
```
Namespace(infinite=True, num_workers=0)
rank: 0, epoch 0, batch [00] (0, tensor([12,  9, 11, 13]))
rank: 0, epoch 0, batch [01] (1, tensor([ 2, 15,  4,  7]))
rank: 0, epoch 1, batch [00] (0, tensor([ 5,  6, 11,  7]))
rank: 0, epoch 1, batch [01] (1, tensor([ 1,  9, 10, 13]))
rank: 0, epoch 2, batch [00] (0, tensor([8, 1, 6, 0]))
rank: 0, epoch 2, batch [01] (1, tensor([ 4, 10, 15,  5]))
rank: 1, epoch 0, batch [00] (0, tensor([10,  6,  8,  5]))
rank: 1, epoch 0, batch [01] (1, tensor([14,  0,  3,  1]))
rank: 1, epoch 1, batch [00] (0, tensor([15,  4,  2, 12]))
rank: 1, epoch 1, batch [01] (1, tensor([ 0,  8,  3, 14]))
rank: 1, epoch 2, batch [00] (0, tensor([13,  9,  7, 11]))
rank: 1, epoch 2, batch [01] (1, tensor([ 2, 12, 14,  3]))
```

When called like this:
```
CUDA_VISIBLE_DEVICES=2,3 python3 -m torch.distributed.run --nproc_per_node 2 test_distributed.py --num-workers=4
```
it outputs:
```
Namespace(infinite=False, num_workers=4)
rank: 0, epoch 0, batch [00] (0, tensor([12,  9, 11, 13]))
rank: 0, epoch 0, batch [01] (1, tensor([ 2, 15,  4,  7]))
rank: 0, epoch 1, batch [00] (0, tensor([ 5,  6, 11,  7]))
rank: 0, epoch 1, batch [01] (1, tensor([ 1,  9, 10, 13]))
rank: 0, epoch 2, batch [00] (0, tensor([8, 1, 6, 0]))
rank: 0, epoch 2, batch [01] (1, tensor([ 4, 10, 15,  5]))
rank: 1, epoch 0, batch [00] (0, tensor([10,  6,  8,  5]))
rank: 1, epoch 0, batch [01] (1, tensor([14,  0,  3,  1]))
rank: 1, epoch 1, batch [00] (0, tensor([15,  4,  2, 12]))
rank: 1, epoch 1, batch [01] (1, tensor([ 0,  8,  3, 14]))
rank: 1, epoch 2, batch [00] (0, tensor([13,  9,  7, 11]))
rank: 1, epoch 2, batch [01] (1, tensor([ 2, 12, 14,  3]))
```

When called like this:
```
CUDA_VISIBLE_DEVICES=2,3 python3 -m torch.distributed.run --nproc_per_node 2 test_distributed.py --num-workers=0
```
it outputs:
```
Namespace(infinite=False, num_workers=0)
rank: 0, epoch 0, batch [00] (0, tensor([12,  9, 11, 13]))
rank: 0, epoch 0, batch [01] (1, tensor([ 2, 15,  4,  7]))
rank: 0, epoch 1, batch [00] (0, tensor([ 5,  6, 11,  7]))
rank: 0, epoch 1, batch [01] (1, tensor([ 1,  9, 10, 13]))
rank: 0, epoch 2, batch [00] (0, tensor([8, 1, 6, 0]))
rank: 0, epoch 2, batch [01] (1, tensor([ 4, 10, 15,  5]))
rank: 1, epoch 0, batch [00] (0, tensor([10,  6,  8,  5]))
rank: 1, epoch 0, batch [01] (1, tensor([14,  0,  3,  1]))
rank: 1, epoch 1, batch [00] (0, tensor([15,  4,  2, 12]))
rank: 1, epoch 1, batch [01] (1, tensor([ 0,  8,  3, 14]))
rank: 1, epoch 2, batch [00] (0, tensor([13,  9,  7, 11]))
rank: 1, epoch 2, batch [01] (1, tensor([ 2, 12, 14,  3]))
```

### Environment

- YOLO: YOLOV5 git commit 8efe97719c9b3b77c9db9b5c8592e051b7f0c9a7
- OS Ubuntu 20.04
- Python 3.8
- torch==1.10.2+cu113
- torchvision==0.11.3+cu113

### Minimal Reproducible Example

See example program above. Also attached as a TXT file
[test_distributed.py.txt](https://github.com/ultralytics/yolov5/files/8116959/test_distributed.py.txt)

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by glenn-jocher**
Proposed DDP shuffle fix in https://github.com/ultralytics/yolov5/issues/6741

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Removal of InfiniteDataLoader in favor of PyTorch's persistent workers.

### 📊 Key Changes
- Deleted the `InfiniteDataLoader` class, which was a custom dataloader designed to reuse workers.
- Removed the internal `_RepeatSampler` class used by the `InfiniteDataLoader`.
- Modified the main data loading logic to use PyTorch's built-in `DataLoader` with added `persistent_workers=True` parameter for worker reusability.

### 🎯 Purpose & Impact
- **Purpose**: To streamline the data loading process by utilizing PyTorch's native capabilities for greater efficiency and code maintainability.
- **Impact**: Users should experience more robust and potentially faster data loading due to the optimized use of PyTorch's `DataLoader` with persistent workers, reducing the need for custom classes and simplifying the codebase. This change can also lead to better performance and easier maintenance moving forward. 🚀

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@willyd I ran a quick speed test on Colab+ high-mem. Just checking single-GPU speeds for now. It seems the InfiniteDataloader is providing consistently improved speeds (at least on small datasets like COCO128).
```
!python train.py --img 640 --batch 16 --epochs 100 --data coco128.yaml --weights yolov5s.pt --cache
```
YOLOv5 🚀 v6.1-21-ge6e36aa torch 1.10.0+cu111 CUDA:0 (Tesla V100-SXM2-16GB, 16160MiB)
Setup complete ✅ (8 CPUs, 51.0 GB RAM, 44.8/166.8 GB disk)

### Run 1

&nbsp; |time (hours)    |mAP
---        |---          |--- 
master |0.054  | 0.770
PR         |0.067  | 0.781

### Run 2

&nbsp; |time (hours)    |mAP
---        |---          |--- 
master |0.056  | 0.774
PR         |0.067  | 0.773

**Comment by willyd** [comment]
@glenn-jocher I suppose that if there is more work being done than spawning new workers at each epoch it may be more apparent for smaller datasets. Let me make some tests in DDP mode and get back to you.

**Comment by willyd** [comment]
@glenn-jocher I ran the same command on 4 RTX 2080 Ti GPUs and got:

  | time (hours) | mAP
-- | -- | --
master | 0.062 | 72.4
PR | 0.069 | 73.5

~~So there is a slow down in DDP mode too.~~

EDIT: This was in DP mode actually.

**Comment by willyd** [comment]
@glenn-jocher I ran in DDP mode this time:

  | time (hours) | mAP
-- | -- | --
master | 0.044 | 73.1
PR | 0.047 | 70.9

It appears that the InfiniteDataLoader is consistently faster.

**Comment by glenn-jocher** [comment]
@willyd thanks for the results! Yes the speeds appear consistently faster with master. The mAP variation (PR vs master) is probably just due to noise, though the Multi-GPU mAP drop (vs Single-GPU) is probably due to not using --sync.

I'll go ahead and close this. Even though we weren't able to improve master it's useful to see that the InfiniteDataloader is providing benefits.

**Comment by willyd** [comment]
@glenn-jocher Thanks for mentionning about the `--sync` option I will keep this in mind.

I still can't totally understand why my test script in #6741 had shuffling problems and YOLOv5 does not. To get a better understanding I looked at the source code and tried to step through the code and found that contrarily to PyTorch's DataLoader which calls `_reset` on the iterator on each `__iter__` call the `InfiniteDataLoader` does not.

See this: https://github.com/pytorch/pytorch/blob/v1.7.0/torch/utils/data/dataloader.py#L349

If I add the call in `InfiniteDataLoader.__iter__` I don't get repetitions with that test script.

```python
class InfiniteDataLoader(dataloader.DataLoader):
    """ Dataloader that reuses workers

    Uses same syntax as vanilla DataLoader
    """

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        object.__setattr__(self, 'batch_sampler', _RepeatSampler(self.batch_sampler))
        self.iterator = super().__iter__()

    def __len__(self):
        return len(self.batch_sampler.sampler)

    def __iter__(self):
        # PyTorch resets the iterator when persistent_workers==True
        self.iterator._reset(self)
        for i in range(len(self)):
            yield next(self.iterator)
```
Although, there does not seem to be a problem with shuffling in the case of YOLOv5 there could be other side effects we did not notice. Even if it increases training time, I think it would be safer to stick with the official implementation and merge this PR or keep investigating the issue.

What's your take on this?

**Comment by willyd** [comment]
Some more information. With the current implementation of `InfiniteDataLoader` my test script repeats the same batch for the first few epochs (depending on batch size and dataset size) and then the batches do not repeat themselves.

**Comment by glenn-jocher** [comment]
@willyd tested again on VOC 10 epochs. Slowdown seems to mainly affect epoch switchover, so for larger datasets slowdown seems to be less of an issue.

```python
!git checkout master
!python train.py --img 512 --batch 16 --epochs 10 --data VOC.yaml --weights yolov5s.pt --cache
!git checkout update/dataloader
!python train.py --img 512 --batch 16 --epochs 10 --data VOC.yaml --weights yolov5s.pt --cache
```

  | time (hours) | mAP
-- | -- | --
master | 0.322 | 0.545
PR | 0.322 | 0.543

**Comment by glenn-jocher** [comment]
Additional VOC tests (3 runs):

```python
!git checkout master
!python train.py --batch 32 --weights yolov5s.pt --data VOC.yaml --epochs 50 --cache --img 512 --nosave --hyp hyp.VOC.yaml
!git checkout update/dataloader
!python train.py --batch 32 --weights yolov5s.pt --data VOC.yaml --epochs 50 --cache --img 512 --nosave --hyp hyp.VOC.yaml
```

  | time (hours) | mAP
-- | -- | --
master | 1.158 | 0.624
PR | 1.158 | 0.624

  | time (hours) | mAP
-- | -- | --
master | 1.158 | 0.624
PR | 1.158 | 0.624

  | time (hours) | mAP
-- | -- | --
master | 1.146 | 0.623
PR | 1.165 | 0.623

**Comment by willyd** [comment]
@glenn-jocher this is encouraging. Let me know if I can help by performing some tests.

**Comment by glenn-jocher** [comment]
Thanks for the offer, @willyd! If you'd like to run tests, here are a couple of possibilities:
1. Test larger datasets for consistency - e.g. COCO, Open Images or similar.
2. Evaluate training performance differences over 300-600 epochs with or without the `InfiniteDataLoader`.

Any of these would be very enlightening. Let me know if you have any questions!
